### PR TITLE
fix: task ID of image blobs for dragonfly's issue #4418

### PR DESCRIPTION
**Pull request by vulyon**
# Calculate task ID based on blob SHA256 for registry downloads

## Description

This PR implements a feature to calculate task IDs based on the blob's SHA256 digest for OCI registry blob downloads, rather than using the full download URL. This allows identical blobs from different registry domains to share the same task ID, avoiding redundant downloads and storage.

### Key Changes:

1. **Configuration Addition** (`dragonfly-client-config/src/dfdaemon.rs`):
   - Added `enable_blob_based_task_id` field to `RegistryMirror` configuration
   - Defaults to `true` as specified in the issue
   - Allows users to control whether this feature is enabled

2. **Core Implementation** (`dragonfly-client/src/proxy/mod.rs`):
   - Added `extract_blob_digest_from_url()` function to detect and extract SHA256 digest from registry blob URLs (format: `/v2/<name>/blobs/sha256:<digest>`)
   - Modified `make_download_task_request()` to use blob digest for task ID calculation when the feature is enabled
   - Implements a priority system for task ID calculation:
     1. Explicit `X-Dragonfly-Content-For-Calculating-Task-ID` header (highest priority)
     2. Blob SHA256 digest from URL (if `enable_blob_based_task_id` is true)
     3. URL-based calculation (default fallback)

3. **Testing**:
   - Added unit tests for `extract_blob_digest_from_url()` function
   - Tests cover valid blob URLs and invalid/edge cases

### Example:

Before this change:
- `https://registry1.io/v2/nginx/blobs/sha256:abc123...` → Task ID 1
- `https://registry2.io/v2/nginx/blobs/sha256:abc123...` → Task ID 2 (duplicate download)

After this change:
- Both URLs → Same Task ID (based on `sha256:abc123...`)
- No duplicate downloads, shared storage

## Related Issue

Closes https://github.com/dragonflyoss/dragonfly/issues/4418

## Motivation and Context

When pulling container images, different registry mirrors (e.g., Docker Hub, Harbor, custom registries) may serve the same image blobs with identical content but different URLs. Currently, Dragonfly computes task IDs based on the download URL, which leads to:

- **Multiple downloads** of the same blob from different registries
- **Data redundancy** in storage
- **Wasted bandwidth** and slower image pulls

This change solves the problem by:
- Using the blob's SHA256 digest (which is content-addressable) as the basis for task ID calculation
- Ensuring identical blobs share the same task ID regardless of registry domain
- Reducing storage usage and improving download efficiency

The feature is **enabled by default** but can be disabled via configuration if needed for compatibility reasons.

## Screenshots (if appropriate)

N/A - Backend functionality only

### CODE REVIEW

File: dragonfly-client-config/src/dfdaemon.rs [hunk 1]
```diff
@@ -1120,6 +1126,14 @@ pub struct RegistryMirror {
...
     /// cert for the registry mirror.
     pub cert: Option<PathBuf>,
+
+    /// enable_blob_based_task_id indicates whether to calculate the task ID based on the blob's SHA256 digest
+    /// for OCI registry blob download URLs. When enabled, if the download URL is for an image blob
+    /// (e.g., /v2/<name>/blobs/sha256:<digest>), the task ID will be calculated based on the blob's digest
+    /// instead of the full URL. This allows the same blob from different registries to share the same task ID,
+    /// avoiding redundant downloads and storage.
+    #[serde(default = "default_proxy_registry_mirror_enable_blob_based_task_id")]
+    pub enable_blob_based_task_id: bool,
```
Comment (gaius-qi) [inline dragonfly-client-config/src/dfdaemon.rs]: Rename to `enable_task_id_based_blob_digest`.

Comment (vulyon) [reply dragonfly-client-config/src/dfdaemon.rs]: ok
